// @atlaskit/app-provider/dist/cjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "UNSAFE_useColorModeForMigration", {
  enumerable: true,
  get: function get() {
    return _themeProvider.UNSAFE_useColorModeForMigration;
  }
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _appProvider.default;
  }
});
Object.defineProperty(exports, "useColorMode", {
  enumerable: true,
  get: function get() {
    return _themeProvider.useColorMode;
  }
});
Object.defineProperty(exports, "useRouterLink", {
  enumerable: true,
  get: function get() {
    return _useRouterLink.default;
  }
});
Object.defineProperty(exports, "useSetColorMode", {
  enumerable: true,
  get: function get() {
    return _themeProvider.useSetColorMode;
  }
});
Object.defineProperty(exports, "useSetTheme", {
  enumerable: true,
  get: function get() {
    return _themeProvider.useSetTheme;
  }
});
Object.defineProperty(exports, "useTheme", {
  enumerable: true,
  get: function get() {
    return _themeProvider.useTheme;
  }
});
var _appProvider = _interopRequireDefault(require("./app-provider"));
var _themeProvider = require("./theme-provider");
var _useRouterLink = _interopRequireDefault(require("./router-link-provider/hooks/use-router-link"));
