// react-magnetic-di/lib/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DiProvider", {
  enumerable: true,
  get: function get() {
    return _provider.DiProvider;
  }
});
Object.defineProperty(exports, "debug", {
  enumerable: true,
  get: function get() {
    return _utils.debug;
  }
});
Object.defineProperty(exports, "di", {
  enumerable: true,
  get: function get() {
    return _consumer.di;
  }
});
Object.defineProperty(exports, "injectable", {
  enumerable: true,
  get: function get() {
    return _injectable.injectable;
  }
});
Object.defineProperty(exports, "runWithDi", {
  enumerable: true,
  get: function get() {
    return _global.runWithDi;
  }
});
Object.defineProperty(exports, "stats", {
  enumerable: true,
  get: function get() {
    return _stats.stats;
  }
});
Object.defineProperty(exports, "unstable_globalDi", {
  enumerable: true,
  get: function get() {
    return _global.globalDi;
  }
});
Object.defineProperty(exports, "withDi", {
  enumerable: true,
  get: function get() {
    return _provider.withDi;
  }
});
var _consumer = require("./react/consumer");
var _global = require("./react/global");
var _provider = require("./react/provider");
var _injectable = require("./react/injectable");
var _utils = require("./react/utils");
var _stats = require("./react/stats");
