// @atlaskit/primitives/dist/cjs/components/metric-text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = require("react");
var _react2 = require("@emotion/react");
var _tinyInvariant = _interopRequireDefault(require("tiny-invariant"));
var _styleMaps = require("../xcss/style-maps.partial");
/**
 * @jsxRuntime classic
 * @jsx jsx
 */

// eslint-disable-next-line @atlaskit/ui-styling-standard/use-compiled -- Ignored via go/DSP-18766

var asAllowlist = ['span', 'div'];
// We're doing this because our CSS reset can add top margins to elements such as `p` which is totally insane.
// Long term we should remove those instances from the reset - it should be a reset to 0.
// For now, at least we know <MetricText> will be unaffected by this.
var resetStyles = (0, _react2.css)({
  margin: 0
});
var textAlignMap = {
  center: (0, _react2.css)({
    textAlign: 'center'
  }),
  end: (0, _react2.css)({
    textAlign: 'end'
  }),
  start: (0, _react2.css)({
    textAlign: 'start'
  })
};

/**
 * __MetricText__
 *
 * MetricText is a primitive component that has the Atlassian Design System's design guidelines baked in.
 * It is designed for use specifically with displaying metrics and is not to be used for headings or general UI text.
 * It renders a `span` by default.
 *
 * @internal
 */
var MetricText = /*#__PURE__*/(0, _react.forwardRef)(function (_ref, ref) {
  var _ref$as = _ref.as,
    Component = _ref$as === void 0 ? 'span' : _ref$as,
    align = _ref.align,
    testId = _ref.testId,
    id = _ref.id,
    size = _ref.size,
    children = _ref.children;
  (0, _tinyInvariant.default)(asAllowlist.includes(Component), "@atlaskit/primitives: MetricText received an invalid \"as\" value of \"".concat(Component, "\""));
  var component = (0, _react2.jsx)(Component, {
    ref: ref,
    css: [resetStyles,
    // eslint-disable-next-line @atlaskit/ui-styling-standard/no-imported-style-values
    size && _styleMaps.metricTextSizeStylesMap[size], align && textAlignMap[align]],
    "data-testid": testId,
    id: id
  }, children);
  return component;
});
var _default = exports.default = MetricText;
