// @atlaskit/platform-feature-flags-react/dist/cjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "componentWithCondition", {
  enumerable: true,
  get: function get() {
    return _componentWithCondition.componentWithCondition;
  }
});
Object.defineProperty(exports, "componentWithFG", {
  enumerable: true,
  get: function get() {
    return _componentWithFg.componentWithFG;
  }
});
Object.defineProperty(exports, "conditionalHooksFactory", {
  enumerable: true,
  get: function get() {
    return _conditionalHooksFactory.conditionalHooksFactory;
  }
});
Object.defineProperty(exports, "functionUnionWithCondition", {
  enumerable: true,
  get: function get() {
    return _functionWithCondition.functionUnionWithCondition;
  }
});
Object.defineProperty(exports, "functionWithCondition", {
  enumerable: true,
  get: function get() {
    return _functionWithCondition.functionWithCondition;
  }
});
Object.defineProperty(exports, "functionWithFG", {
  enumerable: true,
  get: function get() {
    return _functionWithFg.functionWithFG;
  }
});
var _componentWithCondition = require("./utils/component-with-condition");
var _componentWithFg = require("./utils/component-with-fg");
var _functionWithCondition = require("./utils/function-with-condition");
var _functionWithFg = require("./utils/function-with-fg");
var _conditionalHooksFactory = require("./utils/conditional-hooks-factory");
