// @atlaskit/layering/dist/cjs/components/open-layer-observer/use-open-layer-observer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useOpenLayerObserver = useOpenLayerObserver;
var _react = require("react");
var _tinyInvariant = _interopRequireDefault(require("tiny-invariant"));
var _openLayerObserverContext = require("./open-layer-observer-context");
/**
 * A hook for use within an `OpenLayerObserver` component. It provides access to:
 *
 * - `getCount`: a function that returns the current count of open layers under the observer.
 * - `onChange`: a function that allows you to subscribe to changes in the layer count. It returns a
 * cleanup function to unsubscribe, which you should call when the component unmounts.
 * - `closeLayers`: a function that closes all open layers.
 *
 * Example usage:
 * ```tsx
 * const openLayerObserver = useOpenLayerObserver();
 * useEffect(() => {
 *   return openLayerObserver.onChange(function onChange({ count }) {
 *     // react to changes in the layer count
 *   });
 * }, [openLayerObserver]);
 * ```
 */
function useOpenLayerObserver() {
  var context = (0, _react.useContext)(_openLayerObserverContext.OpenLayerObserverContext);
  (0, _tinyInvariant.default)(context, 'useOpenLayerObserver must be used within an OpenLayerObserver');
  var publicAPI = (0, _react.useMemo)(function () {
    return {
      getCount: context.getCount,
      onChange: context.onChange,
      closeLayers: context.closeLayers
    };
  }, [context.getCount, context.onChange, context.closeLayers]);
  return publicAPI;
}
